client/src/main/java/io/pravega/client/tables/TableSegment.java

```diff
@@ -0,0 +1,130 @@
+/**
+ * Copyright (c) 2017 Dell Inc., or its subsidiaries. All Rights Reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ */
+package io.pravega.client.tables;
+
+import io.pravega.common.util.AsyncIterator;
+import java.util.List;
+import java.util.concurrent.CompletableFuture;
+import lombok.Data;
+
+/**
+ * Defines all operations that are supported on a Table Segment.
+ *
+ * @param <KeyT>   Table Key Type.
+ * @param <ValueT> Table Value Type.
+ */
+public interface TableSegment<KeyT, ValueT> extends AutoCloseable {
```

Comment:
As discussed offline, we also need support for `getAllKeys` api on TableSegment. 
This should return all active keys in the segment. Typically this should be a result of a full scan of index (which should typically be small if hard deletes of keys from the index are allowed).

Reply:
You can use the `iterator` for this.

Reply:
If you have a Table with 100M keys (or more), you will not be able to get all of them in one go. Not a chance. We can configure the server-side wire protocol to buffer as much as it can (maybe up to 1M?) before sending over the wire.

Reply:
`getAllKeys` will do exactly what `keyIterator` does (note name change - I now have `keyIterator` and `entryIterator`); it's just a different name.

If you look at what kind of iterator this is, it's not your typical Java iterator. It is an `AsyncIterator` where the iteration item contains a collection of items (keys). If all of your keys will fit in one such batch, then they will all be returned in one call to the server. 

We could change this API to include some sort of hints or configuration where it says what's the maximum number of keys to return at once. Then the caller can specify a large enough value to make this return all keys at once; if the value is too large to be able to ship that data over the wire, then it's the caller's problem to deal with it.